### PR TITLE
Wip/hughsie/native impi

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
FYI: untested as a whole, but I've tested each bit. I'll be able to test this tomorrow.

**Comment by hughsie** [comment]
With a few little fixups this works beautifully on real server hardware.

File: plugins/redfish/fu-ipmi-device.c [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+	if (g_getenv("FWUPD_REDFISH_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "ipmi-recv", buf, *len);
+	if (netfn != NULL)
+		*netfn = recv.msg.netfn;
+	if (cmd != NULL)
+		*cmd = recv.msg.cmd;
+	if (seq != NULL)
+		*seq = recv.msgid;
+	if (len != NULL)
+		*len = (gsize)recv.msg.data_len;
```
Comment (superm1) [inline plugins/redfish/fu-ipmi-device.c:116]: should any of these set an error?

File: plugins/redfish/fu-ipmi-device.c [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+		}
+		fu_device_set_version(device, str->str);
+		bcd = resp[4] & 0x0f;
+		bcd += 10 * (resp[4] >> 4);
+		self->version_ipmi = bcd;
+	} else {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "failed to retrieve Device ID from IPMI");
```
Comment (superm1) [inline plugins/redfish/fu-ipmi-device.c:347]: including resp_length in the error perhaps?

File: plugins/redfish/fu-ipmi-device.c (hunk 1 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:116]: I don't _think_ so. `fu_ipmi_device_recv` just gets the response, no matter the contents.

File: plugins/redfish/fu-ipmi-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:347]: Good idea, fixed.

File: plugins/redfish/fu-ipmi-device.c [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+	FuDeviceClass *klass_device = FU_DEVICE_CLASS(klass);
+	klass_device->setup = fu_ipmi_device_setup;
+	klass_device->to_string = fu_ipmi_device_to_string;
+}
+
+FuIpmiDevice *
+fu_ipmi_device_new(FuContext *ctx)
+{
+	FuIpmiDevice *self;
+	self = g_object_new(FU_TYPE_IPMI_DEVICE, "context", ctx, "device-file", "/dev/ipmi0", NULL);
```
Comment (superm1) [inline plugins/redfish/fu-ipmi-device.c:571]: can't this also be `/dev/ipmidev/0`?  Or are those just symlinks?

Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:571]: I don't see any of `/dev/ipmidev` at all on RHEL/Fedora.

File: plugins/redfish/fu-ipmi-device.c [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_ipmi_device_init(FuIpmiDevice *self)
+{
+	fu_device_set_physical_id(FU_DEVICE(self), "/dev/ipmi0");
```
Comment (superm1) [inline plugins/redfish/fu-ipmi-device.c:552]: can this come from udev instead?  If not, do we need to also make sure that we load `impi-si` and `ipmi-devintf`?

File: plugins/redfish/fu-ipmi-device.c (hunk 1 above)
Comment (superm1) [reply plugins/redfish/fu-ipmi-device.c:116]: Okay.

File: plugins/redfish/fu-ipmi-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:552]: not really; there's no `udev_device` at all in FuUdevDevice, it's literally a hardcoded `device-file` we're using.

File: plugins/redfish/fu-ipmi-device.c (hunk 3 above)
Comment (superm1) [reply plugins/redfish/fu-ipmi-device.c:571]: I don't know the context of it - but I see a few different things tried in ipmitool...
https://github.com/ipmitool/ipmitool/blob/d674bfdc9a2447706ea9b0b943c98bcb0bbe80d2/src/plugins/open/open.c#L102

File: plugins/redfish/fu-ipmi-device.c (hunk 4 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:552]: > make sure that we load impi-si and ipmi-devintf

They seem loaded by default on Fedora and RHEL, is that a downstream thing?

File: plugins/redfish/fu-ipmi-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:571]: I had no idea they existed. I don't think I want to support more than one BMC device (!!!) but I guess it makes sense to search for /dev/ipmi/0 and /dev/ipmidev/0 too

File: plugins/redfish/fu-ipmi-device.c (hunk 4 above)
Comment (superm1) [reply plugins/redfish/fu-ipmi-device.c]: What is `CONFIG_IPMI_DEVICE_INTERFACE` set to?  I don't see anything for modaliases for the ipmi-devintf module in https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/drivers/char/ipmi/ipmi_devintf.c so unless it's compiled into the kernel I would think that module needs to get loaded _somehow_.

Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c]: > What is CONFIG_IPMI_DEVICE_INTERFACE set to

CONFIG_IPMI_DEVICE_INTERFACE=m

Maybe it's safer to manually load impi-si and ipmi-devintf? I'll do just that.

Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c]: Fixed.

File: plugins/redfish/fu-ipmi-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/redfish/fu-ipmi-device.c:591]: Fixed.

**Review by superm1 — DISMISSED** [review]

File: plugins/redfish/fu-plugin-redfish.c [hunk 5]
```diff
@@ -181,6 +185,89 @@ fu_redfish_plugin_discover_smbios_table(FuPlugin *plugin, GError **error)
...
+	}
+	return g_string_free(str, FALSE);
+}
+
+static gboolean
+fu_redfish_plugin_ipmi_create_user(FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(plugin);
+	const gchar *username_fwupd = "fwupd";
+	guint8 user_id = 0x04;
```
Comment (tcap26) [inline plugins/redfish/fu-plugin-redfish.c:208]: looks like you hardcoded the user_id = 0x04 ..... you probably need to add some logic to handle cases where 0x04 is used by other users, where you need to pick another channel number for user_id.

Comment (hughsie) [reply plugins/redfish/fu-plugin-redfish.c:208]: Yup, certainly a valid criticism. I guess the best thing to do at startup is:

 * look from `user_id` 2 to 63 for a user called fwupd
 * if found, report an error as we didn't save the generated password and can't delete the user
 * if not found, create a fwupd user in the first unused slot and continue to enumerate

@tcap26 does that sound [more] sane?

**Pull request merged**